### ISSUE TITLE
[tx] Implement custom loss functions

### ISSUE THREAD

**Issue by pcmoritz** [NovaSky-AI/SkyRL#482]
Implement the forward_backward_custom function from https://tinker-docs.thinkingmachines.ai/losses so we can actually run RL algorithms.

**Comment by pcmoritz** [comment]
Any progress about this would be welcome, e.g.

- Implement the endpoints in the API server
- Figure out how to integrate it into the engine

**Comment by Guido1Alessandro1Trevisan** [comment]
I'm looking into this now.

A few notes:
- I think I should use per_token_losses = jax.lax.switch(...). for the different loss functions
. I’ll also need to implement the /v1/forward endpoint for reward evaluation and value estimation in PPO.
. By default, PPO also requires the /sampler endpoint, which I’ve already implemented the fast api endpoints for but nothing else.

My understanding is that this task overall includes tasks like the /forward endpoint with process_forward_batch, etc.  (more tasks) Is that correct, or am I overestimating the complexity?

**Comment by pcmoritz** [comment]
Great, yes I think you got it exactly right, there are actually two (separate) tasks here:

One is to support the "Basic loss functions" mentioned in the docs -- these can be implemented in the current forward_backward endpoint, and will need a way to have a slightly different computation graph for the loss part, very likely using one of the conditional jax functions is the way to go like you mentioned.

The second one is implementing separate forward and backward endpoints, so the flow can be like: User submits forward, gets the logits, computes the custom loss on the client side, passes it back to the server, which then does the rest of the computation to get the gradients.

Probably the first one is simpler to start with. You should do whichever you are interested in / feel more comfortable with (currently none is being actively worked on, but will probably be soon if you don't want to do it).

**Comment by Guido1Alessandro1Trevisan** [comment]
Here’s the forward_backward attempt (1) https://github.com/NovaSky-AI/SkyRL/pull/498

**Comment by benji-cannot-code** [comment]
I will take a stab at the [flexible custom loss function](https://tinker-docs.thinkingmachines.ai/losses#flexible-loss-functions-forward_backward_custom)!
